File: src/lib/features.ts
/**
 * Feature extraction — the browser-side port of model/fertiluna/features.py.
 *
 * ⚠️ PARITY CONTRACT: this must produce identical output to the Python
 * implementation for the same input. The Python test suite emits
 * `feature-fixtures.json`; `features.test.ts` loads it and asserts equality
 * within float32 tolerance. If you change one side, change both and re-run
 * the parity test.
 *
 * Conventions (identical to Python):
 *   - A missing day is represented by `NaN`, never 0.
 *   - All features are scalars; output is a fixed Float32Array of length N_FEATURES.
 *   - `estimated_ovulation_day` uses the SENSIPLAN 3-over-6 rule.
 */

import { CYCLE_MAX_DAYS, N_FEATURES } from "./constants";

function nanCount(a: Float64Array | number[]): number {
  let c = 0;
  for (let i = 0; i < a.length; i++) if (!Number.isNaN(a[i])) c++;
  return c;
}

function nanMean(a: Float64Array | number[]): number {
  let sum = 0;
  let n = 0;
  for (let i = 0; i < a.length; i++) {
    if (!Number.isNaN(a[i])) {
      sum += a[i];
      n++;
    }
  }
  return n === 0 ? 0 : sum / n;
}

/** Population std (ddof=0), matching numpy's default. Returns 0 if < 2 values. */
function nanStd(a: Float64Array | number[]): number {
  const vals: number[] = [];
  for (let i = 0; i < a.length; i++) if (!Number.isNaN(a[i])) vals.push(a[i]);
  if (vals.length < 2) return 0;
  const mean = vals.reduce((s, v) => s + v, 0) / vals.length;
  let acc = 0;
  for (const v of vals) acc += (v - mean) * (v - mean);
  return Math.sqrt(acc / vals.length);
}

function nanMin(a: Float64Array | number[]): number {
  let m = Infinity;
  for (let i = 0; i < a.length; i++)
    if (!Number.isNaN(a[i]) && a[i] < m) m = a[i];
  return m === Infinity ? 0 : m;
}

function nanMax(a: Float64Array | number[]): number {
  let m = -Infinity;
  for (let i = 0; i < a.length; i++)
    if (!Number.isNaN(a[i]) && a[i] > m) m = a[i];
  return m === -Infinity ? 0 : m;
}

function nanMedian(a: Float64Array | number[]): number {
  const vals: number[] = [];
  for (let i = 0; i < a.length; i++) if (!Number.isNaN(a[i])) vals.push(a[i]);
  if (vals.length === 0) return 0;
  vals.sort((x, y) => x - y);
  const mid = Math.floor(vals.length / 2);
  return vals.length % 2 === 1 ? vals[mid] : (vals[mid - 1] + vals[mid]) / 2;
}

function slice(a: Float64Array, start: number, end: number): Float64Array {
  return a.subarray(Math.max(0, start), Math.min(a.length, end));
}

/**
 * SENSIPLAN 3-over-6 rule. Find the first day d such that temps[d..d+2] are all
 * strictly greater than max(temps[d-6..d-1]), with the 3rd day ≥ 0.20 above that
 * max. Returns the index of the LAST follicular day (one before the rise), or -1.
 */
function detectOvulationDaySensiplan(temps: Float64Array): number {
  const n = temps.length;
  for (let d = 6; d < n - 2; d++) {
    const prior = slice(temps, d - 6, d);
    if (nanCount(prior) < 4) continue;
    const rising = slice(temps, d, d + 3);
    if (nanCount(rising) < 3) continue;
    const priorMax = nanMax(prior);
    let allAbove = true;
    for (let i = 0; i < rising.length; i++) {
      if (!(rising[i] > priorMax)) {
        allAbove = false;
        break;
      }
    }
    if (!allAbove) continue;
    if (rising[2] - priorMax < 0.2) continue;
    return d - 1;
  }
  return -1;
}

function longestRun(mask: boolean[]): number {
  let longest = 0;
  let current = 0;
  for (const v of mask) {
    if (v) {
      current++;
      if (current > longest) longest = current;
    } else {
      current = 0;
    }
  }
  return longest;
}

function findLocalMaxima(arr: Float64Array, threshold: number): number[] {
  const out: number[] = [];
  const n = arr.length;
  for (let i = 0; i < n; i++) {
    const v = arr[i];
    if (Number.isNaN(v) || v <= threshold) continue;
    let left = i > 0 ? arr[i - 1] : -Infinity;
    let right = i < n - 1 ? arr[i + 1] : -Infinity;
    if (Number.isNaN(left)) left = -Infinity;
    if (Number.isNaN(right)) right = -Infinity;
    if (v >= left && v >= right) out.push(i);
  }
  return out;
}

/** Least-squares slope over NaN-filtered (dayIndex, value). 0 if < 2 valid. */
function slope(arr: Float64Array): number {
  const xs: number[] = [];
  const ys: number[] = [];
  for (let i = 0; i < arr.length; i++) {
    if (!Number.isNaN(arr[i])) {
      xs.push(i);
      ys.push(arr[i]);
    }
  }
  if (xs.length < 2) return 0;
  const xMean = xs.reduce((s, v) => s + v, 0) / xs.length;
  const yMean = ys.reduce((s, v) => s + v, 0) / ys.length;
  let num = 0;
  let den = 0;
  for (let i = 0; i < xs.length; i++) {
    num += (xs[i] - xMean) * (ys[i] - yMean);
    den += (xs[i] - xMean) * (xs[i] - xMean);
  }
  if (den === 0) return 0;
  return num / den;
}

/**
 * Extract the fixed-length feature vector from raw curves.
 *
 * @param temps length CYCLE_MAX_DAYS, °C or NaN for missing
 * @param lh    length CYCLE_MAX_DAYS, relative units or NaN for missing
 * @returns Float32Array of length N_FEATURES
 */
export function extractFeatures(
  temps: Float64Array,
  lh: Float64Array,
): Float32Array {
  if (temps.length !== CYCLE_MAX_DAYS || lh.length !== CYCLE_MAX_DAYS) {
    throw new Error(
      `extractFeatures expects arrays of length ${CYCLE_MAX_DAYS}`,
    );
  }
  const f = new Float32Array(N_FEATURES);

  const nTemp = nanCount(temps);
  const nLh = nanCount(lh);

  f[0] = nTemp;
  f[1] = nLh;
  f[2] = 1 - nTemp / CYCLE_MAX_DAYS;
  f[3] = 1 - nLh / CYCLE_MAX_DAYS;

  f[4] = nanMean(temps);
  f[5] = nanStd(temps);
  f[6] = nTemp > 0 ? nanMin(temps) : 0;
  f[7] = nTemp > 0 ? nanMax(temps) : 0;
  f[8] = f[7] - f[6];

  const ov = detectOvulationDaySensiplan(temps);
  f[9] = ov >= 0 ? ov + 1 : 0;

  if (ov >= 0) {
    const follicular = slice(temps, 0, ov + 1);
    const luteal = slice(temps, ov + 1, CYCLE_MAX_DAYS);
    f[10] = nanMean(follicular);
    f[11] = nanStd(follicular);
    f[12] = nanMean(luteal);
    f[13] = nanStd(luteal);
    f[14] = Math.max(0, f[12] - f[10]);

    let steepness = 0;
    const lo = Math.max(0, ov - 1);
    const hi = Math.min(CYCLE_MAX_DAYS - 1, ov + 3);
    for (let d = lo; d < hi; d++) {
      if (!(Number.isNaN(temps[d]) || Number.isNaN(temps[d + 1]))) {
        steepness = Math.max(steepness, temps[d + 1] - temps[d]);
      }
    }
    f[15] = steepness;

    const plateauThreshold = f[10] + 0.15;
    const postVals: number[] = [];
    for (let i = 0; i < luteal.length; i++)
      if (!Number.isNaN(luteal[i])) postVals.push(luteal[i]);

    let plateauCount = 0;
    for (const v of postVals) if (v > plateauThreshold) plateauCount++;
    f[16] = plateauCount;

    const postMaskFull: boolean[] = [];
    for (let i = 0; i < luteal.length; i++) {
      postMaskFull.push(!Number.isNaN(luteal[i]) && luteal[i] > plateauThreshold);
    }
    f[17] = longestRun(postMaskFull);

    const follMax = nanCount(follicular) > 0 ? nanMax(follicular) : 0;
    let dips = 0;
    for (const v of postVals) if (v < follMax) dips++;
    f[18] = dips;

    f[23] = ov + 1;
    let lutealObserved = 0;
    for (let i = 0; i < luteal.length; i++)
      if (!Number.isNaN(luteal[i])) lutealObserved++;
    f[24] = lutealObserved;
  } else {
    f[10] = nanMean(temps);
    f[11] = nanStd(temps);
    f[12] = 0;
    f[13] = 0;
    f[14] = 0;
    f[15] = 0;
    f[16] = 0;
    f[17] = 0;
    f[18] = 0;
    f[23] = 0;
    f[24] = 0;
  }

  // LH features
  if (nLh > 0) {
    const lhBaseline = nanMedian(lh);
    const peaks = findLocalMaxima(lh, Math.max(1.3, lhBaseline * 1.5));
    if (peaks.length > 0) {
      let best = peaks[0];
      for (const p of peaks) if (lh[p] > lh[best]) best = p;
      f[19] = best + 1;
      f[20] = lh[best];
      f[21] = peaks.length;
      f[22] = ov >= 0 ? ov + 1 - (best + 1) : 0;
    } else {
      f[19] = 0;
      f[20] = nanMax(lh);
      f[21] = 0;
      f[22] = 0;
    }
  } else {
    f[19] = 0;
    f[20] = 0;
    f[21] = 0;
    f[22] = 0;
  }

  // Slopes around (estimated) ovulation
  if (ov >= 0) {
    f[25] = slope(slice(temps, Math.max(0, ov - 4), ov + 1));
    f[26] = slope(slice(temps, ov + 1, Math.min(CYCLE_MAX_DAYS, ov + 5)));
  } else {
    f[25] = slope(slice(temps, 0, 10));
    f[26] = slope(slice(temps, 10, 20));
  }

  if (nTemp > 0) {
    const overallMean = f[4];
    let below = 0;
    const aboveMask: boolean[] = [];
    const belowMask: boolean[] = [];
    for (let i = 0; i < temps.length; i++) {
      const valid = !Number.isNaN(temps[i]);
      const isBelow = valid && temps[i] < overallMean;
      const isAbove = valid && temps[i] >= overallMean;
      if (isBelow) below++;
      aboveMask.push(isAbove);
      belowMask.push(isBelow);
    }
    f[27] = below / Math.max(1, nTemp);
    f[28] = longestRun(aboveMask);
    f[29] = longestRun(belowMask);
  } else {
    f[27] = 0;
    f[28] = 0;
    f[29] = 0;
  }

  return f;
}

/**
 * Convert a per-day input (numbers or null for missing) into a NaN-filled
 * Float64Array of length CYCLE_MAX_DAYS, the format extractFeatures expects.
 */
export function toCurveArray(days: (number | null)[]): Float64Array {
  const out = new Float64Array(CYCLE_MAX_DAYS).fill(NaN);
  for (let i = 0; i < Math.min(days.length, CYCLE_MAX_DAYS); i++) {
    const v = days[i];
    if (v !== null && v !== undefined && Number.isFinite(v)) out[i] = v;
  }
  return out;
}

File: model/fertiluna/features.py
"""Feature extraction from raw (temps, lh) curves.

This module is the canonical implementation. The browser-side TypeScript port
(src/lib/features.ts) MUST produce identical outputs for the same input — there
is a parity test in tests/test_feature_parity.py that compares against fixtures.

Design notes:
    - NaN-safe everywhere. A missing day is missing, not zero.
    - All features are scalars (no variable-length outputs) so the model input
      is a fixed (N_FEATURES,) vector.
    - The "estimated ovulation day" feature uses the SENSIPLAN 3-over-6 rule
      (3 consecutive temps strictly above the max of the prior 6), which is the
      same rule the model is implicitly learning. This gives the model a strong
      handcrafted feature alongside the raw aggregates.
"""

from __future__ import annotations

import numpy as np

from .constants import CYCLE_MAX_DAYS, FEATURE_NAMES, N_FEATURES


def _nan_count(a: np.ndarray) -> int:
    return int(np.sum(~np.isnan(a)))


def _safe_mean(a: np.ndarray) -> float:
    if _nan_count(a) == 0:
        return 0.0
    return float(np.nanmean(a))


def _safe_std(a: np.ndarray) -> float:
    if _nan_count(a) < 2:
        return 0.0
    return float(np.nanstd(a, ddof=0))


def _detect_ovulation_day_sensiplan(temps: np.ndarray) -> int:
    """SENSIPLAN 3-over-6 rule.

    Find the first day d such that temps[d:d+3] are all strictly > max(temps[d-6:d]),
    and the third of those temps is at least 0.2 above that 6-day max.
    Returns the index of the LAST follicular day (one before the rise), or -1 if
    no rise is detected.
    """
    n = len(temps)
    for d in range(6, n - 2):
        prior = temps[d - 6 : d]
        if _nan_count(prior) < 4:
            continue
        rising = temps[d : d + 3]
        if _nan_count(rising) < 3:
            continue
        prior_max = float(np.nanmax(prior))
        if not np.all(rising > prior_max):
            continue
        if rising[2] - prior_max < 0.20:
            continue
        return d - 1   # last follicular day (0-indexed)
    return -1


def _longest_run(mask: np.ndarray) -> int:
    """Longest consecutive run of True in a 1D bool mask."""
    longest = 0
    current = 0
    for v in mask:
        if v:
            current += 1
            longest = max(longest, current)
        else:
            current = 0
    return longest


def _find_local_maxima(arr: np.ndarray, threshold: float) -> list[int]:
    """Indices where arr[i] > threshold and arr[i] >= both neighbors (NaN-safe)."""
    out = []
    n = len(arr)
    for i in range(n):
        v = arr[i]
        if np.isnan(v) or v <= threshold:
            continue
        left = arr[i - 1] if i > 0 else -np.inf
        right = arr[i + 1] if i < n - 1 else -np.inf
        left = -np.inf if np.isnan(left) else left
        right = -np.inf if np.isnan(right) else right
        if v >= left and v >= right:
            out.append(i)
    return out


def extract_features(temps: np.ndarray, lh: np.ndarray) -> np.ndarray:
    """Return a (N_FEATURES,) float32 vector.

    Args:
        temps: (CYCLE_MAX_DAYS,) array of temperatures (°C) with NaN for missing
        lh:    (CYCLE_MAX_DAYS,) array of LH values with NaN for missing

    Indices into the output vector are given by constants.FEATURE_NAMES.
    """
    assert temps.shape == (CYCLE_MAX_DAYS,)
    assert lh.shape == (CYCLE_MAX_DAYS,)

    f = np.zeros(N_FEATURES, dtype=np.float32)

    n_temp = _nan_count(temps)
    n_lh = _nan_count(lh)

    f[0] = n_temp
    f[1] = n_lh
    f[2] = 1.0 - n_temp / CYCLE_MAX_DAYS
    f[3] = 1.0 - n_lh / CYCLE_MAX_DAYS

    f[4] = _safe_mean(temps)
    f[5] = _safe_std(temps)
    f[6] = float(np.nanmin(temps)) if n_temp > 0 else 0.0
    f[7] = float(np.nanmax(temps)) if n_temp > 0 else 0.0
    f[8] = f[7] - f[6]

    # Detected ovulation (last follicular day, 0-indexed). +1 for human-friendly day.
    ov = _detect_ovulation_day_sensiplan(temps)
    f[9] = float(ov + 1) if ov >= 0 else 0.0

    if ov >= 0:
        follicular = temps[: ov + 1]
        luteal = temps[ov + 1 :]
        f[10] = _safe_mean(follicular)
        f[11] = _safe_std(follicular)
        f[12] = _safe_mean(luteal)
        f[13] = _safe_std(luteal)
        f[14] = max(0.0, f[12] - f[10])
        # rise steepness: max single-day jump in [ov-1, ov+3]
        steepness = 0.0
        lo = max(0, ov - 1)
        hi = min(CYCLE_MAX_DAYS - 1, ov + 3)
        for d in range(lo, hi):
            if not (np.isnan(temps[d]) or np.isnan(temps[d + 1])):
                steepness = max(steepness, float(temps[d + 1] - temps[d]))
        f[15] = steepness
        # plateau days: count of post-ovulation days above follicular_mean + 0.15
        plateau_threshold = f[10] + 0.15
        post = luteal[~np.isnan(luteal)]
        f[16] = float(np.sum(post > plateau_threshold))
        # longest consecutive run above plateau threshold
        post_mask_full = ~np.isnan(luteal) & (luteal > plateau_threshold)
        f[17] = _longest_run(post_mask_full)
        # post-rise dips (luteal days dropping below follicular max)
        foll_max = float(np.nanmax(follicular)) if _nan_count(follicular) > 0 else 0.0
        dips = post < foll_max
        f[18] = float(np.sum(dips))
        f[23] = float(ov + 1)
        f[24] = float(np.sum(~np.isnan(luteal)))
    else:
        # No detected rise → leave most fields at 0; report descriptive aggregates only.
        f[10] = _safe_mean(temps)
        f[11] = _safe_std(temps)
        f[12] = 0.0
        f[13] = 0.0
        f[14] = 0.0
        f[15] = 0.0
        f[16] = 0.0
        f[17] = 0.0
        f[18] = 0.0
        f[23] = 0.0
        f[24] = 0.0

    # LH features
    if n_lh > 0:
        lh_baseline = float(np.nanmedian(lh))
        peaks = _find_local_maxima(lh, threshold=max(1.3, lh_baseline * 1.5))
        if peaks:
            # Highest peak
            best = max(peaks, key=lambda i: lh[i])
            f[19] = float(best + 1)        # 1-indexed day
            f[20] = float(lh[best])
            f[21] = float(len(peaks))
            if ov >= 0:
                f[22] = float((ov + 1) - (best + 1))
            else:
                f[22] = 0.0
        else:
            f[19] = 0.0
            f[20] = float(np.nanmax(lh))
            f[21] = 0.0
            f[22] = 0.0
    else:
        f[19] = 0.0
        f[20] = 0.0
        f[21] = 0.0
        f[22] = 0.0

    # Slopes around (estimated) ovulation
    if ov >= 0:
        pre_window = temps[max(0, ov - 4) : ov + 1]
        post_window = temps[ov + 1 : min(CYCLE_MAX_DAYS, ov + 5)]
        f[25] = _slope(pre_window)
        f[26] = _slope(post_window)
    else:
        f[25] = _slope(temps[:10])
        f[26] = _slope(temps[10:20])

    if n_temp > 0:
        overall_mean = f[4]
        valid = ~np.isnan(temps)
        below = (temps < overall_mean) & valid
        above = (temps >= overall_mean) & valid
        f[27] = float(np.sum(below)) / max(1, n_temp)
        f[28] = float(_longest_run(above))
        f[29] = float(_longest_run(below))
    else:
        f[27] = 0.0
        f[28] = 0.0
        f[29] = 0.0

    assert f.shape == (N_FEATURES,)
    return f


def _slope(arr: np.ndarray) -> float:
    """Least-squares slope over NaN-filtered (x, y) where x is day index."""
    valid = ~np.isnan(arr)
    if int(np.sum(valid)) < 2:
        return 0.0
    x = np.arange(len(arr), dtype=np.float64)[valid]
    y = arr[valid].astype(np.float64)
    x_mean = x.mean()
    y_mean = y.mean()
    denom = float(np.sum((x - x_mean) ** 2))
    if denom == 0.0:
        return 0.0
    return float(np.sum((x - x_mean) * (y - y_mean)) / denom)


def batch_extract(temps: np.ndarray, lh: np.ndarray) -> np.ndarray:
    """Apply extract_features to (n, CYCLE_MAX_DAYS) inputs. Returns (n, N_FEATURES)."""
    assert temps.ndim == 2 and lh.ndim == 2
    n = temps.shape[0]
    out = np.empty((n, N_FEATURES), dtype=np.float32)
    for i in range(n):
        out[i] = extract_features(temps[i], lh[i])
    return out


# Re-export names so callers don't have to import constants separately.
__all__ = ["extract_features", "batch_extract", "FEATURE_NAMES", "N_FEATURES"]
